src/irmin-pack/unix/ranges.ml

```diff
@@ -29,44 +29,97 @@ module Stack = struct
+        s.arr.(i + 1) <- len;
         t
     | _ -> push_pair ~off ~len (make t)
 
-  let rec iter_pair fn = function
-    | Empty -> ()
+  let rec to_seq t () =
+    match t with
+    | Empty -> Seq.Nil
     | Stack { len; arr; prev } ->
         assert (len mod 2 = 0);
-        for i = (len / 2) - 1 downto 0 do
-          let off = arr.(2 * i) in
-          let len = arr.((2 * i) + 1) in
-          fn ~off ~len
-        done;
-        iter_pair fn prev
+        let rec go i () =
+          if i < 0 then to_seq prev ()
+          else
+            let entry = (arr.(2 * i), arr.((2 * i) + 1)) in
+            Seq.Cons (entry, go (i - 1))
+        in
+        go ((len / 2) - 1) ()
 end
 
+type entry = int63 * int63
+
 type t = {
-  mutable last : (int63 * int) option;
+  mutable last : entry option;
   mutable ranges : Stack.t;
   mutable count : int;
+  mutable out_of_order : entry list;
 }
 
-let make () = { last = None; ranges = Stack.Empty; count = 0 }
+let make () =
+  { last = None; ranges = Stack.Empty; count = 0; out_of_order = [] }
+
 let count t = t.count
 
 let add ~off ~len t =
   t.count <- t.count + 1;
-  let off_end = Int63.(Syntax.(off + of_int len)) in
+  let open Int63.Syntax in
+  let len = Int63.of_int len in
   match t.last with
   | None -> t.last <- Some (off, len)
-  | Some (off', len') when off_end = off' -> t.last <- Some (off, len + len')
+  | Some (off', len') when off + len = off' -> t.last <- Some (off, len + len')
+  | Some (off', _) when off > off' ->
+      t.out_of_order <- (off, len) :: t.out_of_order
   | Some (off', len') ->
       t.last <- Some (off, len);
       t.ranges <- Stack.push_pair ~off:off' ~len:len' t.ranges
 
-let iter fn t =
+let ranges_to_seq t () =
   match t.last with
-  | None -> assert (t.count = 0)
-  | Some (off, len) ->
-      fn ~off ~len:(Int63.of_int len);
-      Stack.iter_pair fn t.ranges
+  | None -> Seq.Nil
+  | Some entry -> Seq.Cons (entry, Stack.to_seq t.ranges)
+
+let out_of_order_to_seq t =
+  List.to_seq
+  @@ List.sort_uniq
+       (fun (off, _) (off', _) -> Int63.compare off off')
+       t.out_of_order
+
+let rec seq_merge xs ys () =
+  match (xs (), ys ()) with
+  | Seq.Nil, rest | rest, Seq.Nil -> rest
+  | Seq.Cons (((off, len) as x), xs'), Seq.Cons (((off', len') as y), ys') -> (
+      match Int63.compare off off' with
+      | 0 ->
+          assert (len = len');
+          Seq.Cons (x, seq_merge xs' ys')
+      | c when c < 0 -> Seq.Cons (x, seq_merge xs' ys)
+      | _ -> Seq.Cons (y, seq_merge xs ys'))
+
+type fused = Disjoint of entry * entry | Overlap of entry
+
+let fuse ((off, len) as fst) ((off', len') as snd) =
+  let open Int63.Syntax in
+  let fst_end = off + len in
+  let snd_end = off' + len' in
+  if fst_end < off' then Disjoint (fst, snd)
+  else if snd_end < off then Disjoint (snd, fst)
+  else
+    let start = min off off' in
+    let stop = max fst_end snd_end in
+    Overlap (start, stop - start)
+
+let rec seq_fuse ?prev s () =
```

Comment:
I think this fusing is a new idea (?). Do you think it is worth the cost of looping through the entries again in the case that the `out_of_order` list is empty? In the non out of order case, is the benefit of fusing that we could get a smaller mapping file? I don't have an intuition how much benefit we would see.